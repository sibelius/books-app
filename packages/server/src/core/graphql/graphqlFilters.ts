// inspired by: https://github.com/entria/graphql-mongo-helpers

import { ConnectionArguments } from 'graphql-relay';
import idx from 'idx';

import { JSObject } from '@booksapp/types';

import { GraphQLContext } from '../../types';
import { getObjectId } from '../../common/utils';

import { DIRECTION } from './enum/DirectionEnumType';

export const FILTER_CONDITION_TYPE = {
  MATCH_1_TO_1: 'MATCH_1_TO_1', // something that could be used on find() or $match
  CUSTOM_CONDITION: 'CUSTOM_CONDITION', // create a custom condition based on value
  AGGREGATE_PIPELINE: 'AGGREGATE_PIPELINE', // just an aggregate
};

export type BuiltConditionSet = {
  conditions: JSObject;
  pipeline: JSObject[];
};

type FilterFieldMappingMatch<ValueT> = {
  type: typeof FILTER_CONDITION_TYPE.MATCH_1_TO_1;
  key: string;
  format?: (value: ValueT, filter: JSObject, context: GraphQLContext, options?: JSObject<any>) => any;
};

type FilterFieldMappingPipeline<ValueT> = {
  type: typeof FILTER_CONDITION_TYPE.AGGREGATE_PIPELINE;
  pipeline: JSObject[] | ((value: ValueT) => Array<{}>);
};

type FilterFieldMappingCustomCondition<ValueT> = {
  type: typeof FILTER_CONDITION_TYPE.CUSTOM_CONDITION;
  format: (value: ValueT, filter: JSObject, context: GraphQLContext, options?: JSObject<any>) => any;
};

type FilterFieldMapping<ValueT> =
  | FilterFieldMappingMatch<ValueT>
  | FilterFieldMappingPipeline<ValueT>
  | FilterFieldMappingCustomCondition<ValueT>
  | boolean;

export type FilterMapping = { [key: string]: FilterFieldMapping<any> };

const validOperators = ['gt', 'gte', 'lt', 'lte', 'in', 'nin', 'ne', 'all'];

const arrayOperators = ['in', 'nin', 'all'];

const getFilterName = (filterName: string) => filterName.split('_')[0];

type Operators = 'gt' | 'gte' | 'lt' | 'lte' | 'in' | 'nin' | 'ne' | 'all' | '$and' | '$or';

const handleAndOr = (operator: Operators) => (
  context: GraphQLContext,
  condition: Array<{}>,
  mapping: { [key: string]: FilterFieldMapping<{}> },
) => {
  if (!Array.isArray(condition)) {
    throw new Error(`Invalid filter supplied to ${operator}.`);
  }

  return {
    condition: condition.map((andCondition) => buildConditionsObject(context, andCondition, mapping)),
    conditionName: operator,
  };
};

const handleAnd = handleAndOr('$and');
const handleOr = handleAndOr('$or');

const handleFieldOperator = (
  context: GraphQLContext,
  condition: JSObject,
  conditionName: string,
  fieldMapping: FilterFieldMapping<JSObject>,
  prev: JSObject,
  filters: JSObject,
  options = {},
) => {
  // { "myField_operator": "something" } becomes { "myField": { $operator: "something" } }
  // { "myField": "something" } remains the same
  const conditionNamePieces = conditionName.split('_');
  const operator = conditionNamePieces.length > 1 ? conditionNamePieces.pop() : '';
  // I don't think we support snake case for field names, should this be here?
  conditionName = conditionNamePieces.join('_');

  if (fieldMapping && fieldMapping.format && typeof fieldMapping.format === 'function') {
    condition = fieldMapping.format(condition, filters, context, options);
  }

  if (operator) {
    if (validOperators.indexOf(operator) === -1) {
      throw new Error(`"${operator}" is not a valid operator on field "${conditionName}".`);
    }

    if (arrayOperators.indexOf(operator) >= 0 && !Array.isArray(condition)) {
      throw new Error(`Field "${conditionName}" must have an array value.`);
    }

    condition = {
      [`$${operator}`]: condition,
    };
  }

  // handle $gte and $let fields merge
  if (conditionName in prev) {
    condition = {
      ...condition,
      ...prev[conditionName],
    };
  }

  return {
    condition,
    conditionName,
  };
};

function buildConditionsObject<ValueT>(
  context: GraphQLContext,
  conditions: JSObject,
  mapping: { [key: string]: FilterFieldMapping<ValueT> },
  filters?: JSObject,
  options = {},
) {
  return Object.keys(conditions).reduce((prev, currentKey) => {
    let condition = conditions[currentKey];
    let conditionName = currentKey;

    const fieldMapping = mapping[getFilterName(currentKey)];

    if (fieldMapping === false) {
      return prev;
    }

    if (
      fieldMapping &&
      fieldMapping.type !== FILTER_CONDITION_TYPE.MATCH_1_TO_1 &&
      fieldMapping.type !== FILTER_CONDITION_TYPE.CUSTOM_CONDITION
    ) {
      return prev;
    }

    if (conditionName === 'AND') {
      ({ condition, conditionName } = handleAnd(context, condition, mapping));
    } else if (conditionName === 'OR') {
      ({ condition, conditionName } = handleOr(context, condition, mapping));
    } else {
      if (fieldMapping && fieldMapping.type === FILTER_CONDITION_TYPE.CUSTOM_CONDITION) {
        if (fieldMapping.format && typeof fieldMapping.format === 'function') {
          const customCondition = fieldMapping.format(condition, filters, context, options);

          return {
            ...prev,
            ...customCondition,
          };
        }
      }
      ({ condition, conditionName } = handleFieldOperator(
        context,
        condition,
        conditionName,
        fieldMapping,
        prev,
        filters,
        options,
      ));
    }

    conditionName = fieldMapping && fieldMapping.key ? fieldMapping.key : conditionName;

    if (condition === undefined) {
      return prev;
    }

    return {
      ...prev,
      [conditionName]: condition,
    };
  }, {});
}

export function buildMongoConditionsFromFilters<ValueT>(
  context: GraphQLContext,
  filters: JSObject,
  mapping: { [key: string]: FilterFieldMapping<ValueT> } = {},
  options = {},
): BuiltConditionSet {
  if (!filters) {
    return { conditions: {}, pipeline: [] };
  }

  const keys = Object.keys(filters);

  // first check if there are any pipeline mapped fields
  // and if AND or OR are also passed, if that is the case, we must throw an error
  // because we cannot use OR/AND while also using pipeline.
  const hasPipelineFilter = keys.find(
    (item) => idx(mapping[item], (_) => _.type) === FILTER_CONDITION_TYPE.AGGREGATE_PIPELINE,
  );
  if (hasPipelineFilter && (filters.AND || filters.OR)) {
    throw new Error(
      `Wrong filter usage, because filter "${hasPipelineFilter}" is a pipeline filter, which should disable AND and OR`,
    );
  }

  // separate filters by type
  const filtersKeysGrouped = Object.keys(filters).reduce(
    (prev, key) => {
      const filterName = getFilterName(key);

      const type = (mapping && idx(mapping[filterName], (_) => _.type)) || FILTER_CONDITION_TYPE.MATCH_1_TO_1;

      return {
        ...prev,
        [type]: {
          ...prev[type],
          [key]: filters[key],
        },
      };
    },
    {
      // start with sane defaults
      [FILTER_CONDITION_TYPE.MATCH_1_TO_1]: {},
      [FILTER_CONDITION_TYPE.CUSTOM_CONDITION]: {},
      [FILTER_CONDITION_TYPE.AGGREGATE_PIPELINE]: {},
    },
  );

  // first build our conditions JSObject.
  const conditions = buildConditionsObject(
    context,
    { ...filtersKeysGrouped.MATCH_1_TO_1, ...filtersKeysGrouped.CUSTOM_CONDITION },
    mapping,
    filters,
    options,
  );

  // now build the pipeline, which is more straightforward
  const pipeline = Object.keys(filtersKeysGrouped.AGGREGATE_PIPELINE).reduce((prev, key) => {
    const mappedFilter = mapping[key];
    // should not really happen!
    if (!mappedFilter || mappedFilter.type !== FILTER_CONDITION_TYPE.AGGREGATE_PIPELINE) {
      return prev;
    }

    const fieldPipeline = Array.isArray(mappedFilter.pipeline)
      ? mappedFilter.pipeline
      : mappedFilter.pipeline(filters[key], filters, context, options);

    return [...prev, ...(Array.isArray(fieldPipeline) ? fieldPipeline : [])];
  }, []);

  return {
    conditions,
    pipeline,
  };
}

type OrderByArg<SortT> = {
  sort: SortT;
  direction: typeof DIRECTION;
};

export function buildSortFromOrderByArg<Sort>(orderByArg: Array<OrderByArg<string>>): { [key: string]: 1 | -1 } {
  return orderByArg.reduce(
    (acc, item) => ({
      ...acc,
      [item.sort]: item.direction,
    }),
    {},
  );
}

// @TODO - treat when there's more than one $match
export function orderAggregatePipeline(pipeline: JSObject[]): JSObject[] {
  if (!pipeline.length) {
    return [];
  }

  // @TODO - support $geoWithin as well
  const geoNearStage = pipeline.find((stage) => !!stage.$geoNear);

  let orderedPipeline: JSObject[] = [];
  let conditions = {};
  for (const stage of pipeline) {
    // IMPORTANT: if the stage is geoNear put it as first of the pipeline
    if (!!stage.$geoNear) {
      orderedPipeline = [stage, ...orderedPipeline];
      continue;
    }

    // store the first appearing $match on conditions
    // discard it from the ordered pipeline
    if (!!stage.$match) {
      if (geoNearStage && Object.keys(conditions).length === 0) {
        conditions = { ...conditions, ...stage.$match };
        continue;
      }
    }

    orderedPipeline = [...orderedPipeline, stage];
  }

  // add conditions on query of $geoNear
  if (geoNearStage) {
    geoNearStage.$geoNear.query = { ...(geoNearStage.$geoNear.query || {}), ...conditions };
  }

  return orderedPipeline;
}

type ArgsWithFilter = {
  filters: { [key: string]: string };
} & { [key: string]: string } & ConnectionArguments;
export const withFilter = (args: ArgsWithFilter, filters: JSObject) => {
  return {
    ...args,
    filters: {
      ...(args.filters || {}),
      ...filters,
    },
  };
};

export const getFieldsOneToOne = (fields: string[], JSObjectId = false) => {
  const formatParam = JSObjectId ? { format: (value?: string) => value && getObjectId(value) } : {};

  return fields.reduce(
    (acc, field) => ({
      ...acc,
      [field]: {
        type: FILTER_CONDITION_TYPE.MATCH_1_TO_1,
        key: field,
        ...formatParam,
      },
    }),
    {},
  );
};
