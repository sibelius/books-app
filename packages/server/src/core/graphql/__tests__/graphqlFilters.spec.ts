import { getContext } from '../../../../test/helpers';
import { buildMongoConditionsFromFilters, FILTER_CONDITION_TYPE, orderAggregatePipeline } from '../graphqlFilters';

describe('GraphQL Filters', () => {
  it('should return correct conditions for AND comparison', async () => {
    const context = await getContext();

    const filters = {
      AND: [{ status: 'DISABLED' }, { name: 'Jon' }],
    };

    expect(buildMongoConditionsFromFilters(context, filters)).toMatchSnapshot();
  });

  it('should return correct conditions for OR comparison', async () => {
    const context = await getContext();

    const filters = {
      OR: [{ status: 'DISABLED' }, { name: 'Disabled User' }],
    };

    expect(buildMongoConditionsFromFilters(context, filters)).toMatchSnapshot();
  });

  it('should return correct conditions for nested comparison', async () => {
    const context = await getContext();

    const filters = {
      AND: [
        {
          OR: [{ status: 'DISABLED' }, { name: 'Disabled User' }],
        },
        {
          OR: [{ role: 'ADMIN' }, { name: 'Admin' }],
        },
      ],
    };

    expect(buildMongoConditionsFromFilters(context, filters)).toMatchSnapshot();
  });

  it('should throw an error for invalid operator', async () => {
    const context = await getContext();

    const filters = {
      status_yay: 'nay',
    };

    expect(() => {
      buildMongoConditionsFromFilters(context, filters);
    }).toThrowError();
  });

  it('should throw an error for invalid _in and _nin comparison', async () => {
    const context = await getContext();

    const filtersIn = {
      status_in: 'this should be an array',
    };

    expect(() => {
      buildMongoConditionsFromFilters(context, filtersIn);
    }).toThrowError();

    const filtersNin = {
      status_nin: 'this should be an array',
    };

    expect(() => {
      buildMongoConditionsFromFilters(context, filtersNin);
    }).toThrowError();
  });

  it('should return correct conditions for mapped key', async () => {
    const context = await getContext();

    const filters = {
      tags_all: ['a', 'b'],
    };

    const mapping = {
      tags: {
        type: FILTER_CONDITION_TYPE.MATCH_1_TO_1,
        key: 'my.nested.tags',
      },
    };

    expect(buildMongoConditionsFromFilters(context, filters, mapping)).toMatchSnapshot();
  });

  it('should return correct conditions when blacklisting some filters on mapping', async () => {
    const context = await getContext();

    const filters = { a: 'b', c: 'd' };

    const mapping = { c: false };

    expect(buildMongoConditionsFromFilters(context, filters, mapping)).toMatchSnapshot();
  });

  it('should return correct aggregate pipeline', async () => {
    const context = await getContext();

    const filters = { a: 'some-filter-on-a', c: 'this-is-from-another-collection' };

    const mapping = {
      c: {
        type: FILTER_CONDITION_TYPE.AGGREGATE_PIPELINE,
        pipeline: (value: string) => [{ $match: { c: value } }],
      },
    };

    expect(buildMongoConditionsFromFilters(context, filters, mapping)).toMatchSnapshot();
  });

  it('should throw error if using invalid value for filters', async () => {
    const context = await getContext();

    const filtersA = { AND: 'a' };
    expect(() => buildMongoConditionsFromFilters(context, filtersA)).toThrow('Invalid filter supplied to $and.');

    const filtersB = { OR: 'a' };
    expect(() => buildMongoConditionsFromFilters(context, filtersB)).toThrow('Invalid filter supplied to $or.');

    const filtersC = { a: 'a', AND: [] };
    const mappingC = {
      a: {
        type: FILTER_CONDITION_TYPE.AGGREGATE_PIPELINE,
        pipeline: (value: string) => [{ $match: { a: value } }],
      },
    };
    expect(() => buildMongoConditionsFromFilters(context, filtersC, mappingC)).toThrow(
      'Wrong filter usage, because filter "a" is a pipeline filter, which should disable AND and OR',
    );
  });

  it('should return correct conditions when using format function', async () => {
    const context = await getContext();

    const filters = { a: 'something' };

    const mapping = {
      a: {
        type: FILTER_CONDITION_TYPE.MATCH_1_TO_1,
        format: (value: string) => `${value}-else`,
      },
    };

    expect(buildMongoConditionsFromFilters(context, filters, mapping)).toMatchSnapshot();
  });

  it('should work with null filter', async () => {
    const context = await getContext();

    const filters = null;

    expect(buildMongoConditionsFromFilters(context, filters)).toEqual({
      conditions: {},
      pipeline: [],
    });
  });

  it('should work with null condition', async () => {
    const context = await getContext();

    const filters = { a: null };

    expect(buildMongoConditionsFromFilters(context, filters)).toEqual({
      conditions: { a: null },
      pipeline: [],
    });
  });

  it('should work with two dates', async () => {
    const context = await getContext();

    const filters = {
      createdAt_gte: '2018-06-18T17:01:00.000Z',
      createdAt_lte: '2018-07-18T17:01:00.000Z',
    };

    expect(buildMongoConditionsFromFilters(context, filters)).toEqual({
      conditions: {
        createdAt: { $gte: '2018-06-18T17:01:00.000Z', $lte: '2018-07-18T17:01:00.000Z' },
      },
      pipeline: [],
    });
  });

  it('should handle field mapping to a custom condition', async () => {
    const context = await getContext();

    const filters = {
      search: 'something',
    };

    const mapping = {
      search: {
        type: FILTER_CONDITION_TYPE.CUSTOM_CONDITION,
        format: (search: string) => ({
          $or: [{ name: search }, { email: search }],
        }),
      },
    };

    expect(buildMongoConditionsFromFilters(context, filters, mapping)).toEqual({
      conditions: {
        $or: [{ name: 'something' }, { email: 'something' }],
      },
      pipeline: [],
    });
  });

  it('should order a pipeline with geoNear', () => {
    const pipeline = [
      { $match: { something: true } },
      {
        $geoNear: {
          near: { type: 'Point', coordinates: [10.0, 20.0] },
          distanceField: 'dist.calculated',
          includeLocs: 'dist.location',
          spherical: true,
        },
      },
      { $match: { isActive: false } },
    ];

    const orderedPipeline = orderAggregatePipeline(pipeline);
    expect(!!orderedPipeline[0].$geoNear).toBe(true);
  });

  it('should add query to $geoNear when there is $match, only include just match', () => {
    const $match = {
      something: true,
    };

    const pipeline = [
      { $match },
      {
        $geoNear: {
          near: { type: 'Point', coordinates: [10.0, 20.0] },
          distanceField: 'dist.calculated',
          includeLocs: 'dist.location',
          spherical: true,
        },
      },
      { $match: { isActive: false } },
    ];

    const orderedPipeline = orderAggregatePipeline(pipeline);

    expect(!!orderedPipeline[0].$geoNear).toBe(true);
    expect(orderedPipeline[0].$geoNear.query).toEqual($match);
  });

  it('should handle field mapping to a custom condition and options param', async () => {
    const context = await getContext();

    const filters = { search: 'something' };
    const options = { customString: 'customString' };

    const mapping = {
      search: {
        type: FILTER_CONDITION_TYPE.CUSTOM_CONDITION,
        format: (search, filters, mapping, { customString }) => ({
          $or: [{ name: `${search} ${customString}` }, { email: `${search} ${customString}` }],
        }),
      },
    };

    expect(buildMongoConditionsFromFilters(context, filters, mapping, options)).toEqual({
      conditions: {
        $or: [{ name: 'something customString' }, { email: 'something customString' }],
      },
      pipeline: [],
    });
  });

  it('should return correct aggregate pipeline with options param', async () => {
    const context = await getContext();

    const filters = {
      a: 'some-filter-on-a',
      c: 'this-is-from-another-collection',
    };
    const options = { customString: 'customString' };

    const mapping = {
      c: {
        type: FILTER_CONDITION_TYPE.AGGREGATE_PIPELINE,
        pipeline: (value, filters, mapping, { customString }) => [{ $match: { c: `${value} ${customString}` } }],
      },
    };

    expect(buildMongoConditionsFromFilters(context, filters, mapping, options)).toMatchSnapshot();
  });
});
