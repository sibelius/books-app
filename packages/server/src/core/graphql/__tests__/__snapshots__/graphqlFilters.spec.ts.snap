// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL Filters should return correct aggregate pipeline 1`] = `
Object {
  "conditions": Object {
    "a": "some-filter-on-a",
  },
  "pipeline": Array [
    Object {
      "$match": Object {
        "c": "this-is-from-another-collection",
      },
    },
  ],
}
`;

exports[`GraphQL Filters should return correct aggregate pipeline with options param 1`] = `
Object {
  "conditions": Object {
    "a": "some-filter-on-a",
  },
  "pipeline": Array [
    Object {
      "$match": Object {
        "c": "this-is-from-another-collection customString",
      },
    },
  ],
}
`;

exports[`GraphQL Filters should return correct conditions for AND comparison 1`] = `
Object {
  "conditions": Object {
    "$and": Array [
      Object {
        "status": "DISABLED",
      },
      Object {
        "name": "Jon",
      },
    ],
  },
  "pipeline": Array [],
}
`;

exports[`GraphQL Filters should return correct conditions for OR comparison 1`] = `
Object {
  "conditions": Object {
    "$or": Array [
      Object {
        "status": "DISABLED",
      },
      Object {
        "name": "Disabled User",
      },
    ],
  },
  "pipeline": Array [],
}
`;

exports[`GraphQL Filters should return correct conditions for mapped key 1`] = `
Object {
  "conditions": Object {
    "my.nested.tags": Object {
      "$all": Array [
        "a",
        "b",
      ],
    },
  },
  "pipeline": Array [],
}
`;

exports[`GraphQL Filters should return correct conditions for nested comparison 1`] = `
Object {
  "conditions": Object {
    "$and": Array [
      Object {
        "$or": Array [
          Object {
            "status": "DISABLED",
          },
          Object {
            "name": "Disabled User",
          },
        ],
      },
      Object {
        "$or": Array [
          Object {
            "role": "ADMIN",
          },
          Object {
            "name": "Admin",
          },
        ],
      },
    ],
  },
  "pipeline": Array [],
}
`;

exports[`GraphQL Filters should return correct conditions when blacklisting some filters on mapping 1`] = `
Object {
  "conditions": Object {
    "a": "b",
  },
  "pipeline": Array [],
}
`;

exports[`GraphQL Filters should return correct conditions when using format function 1`] = `
Object {
  "conditions": Object {
    "a": "something-else",
  },
  "pipeline": Array [],
}
`;
